### ISSUE TITLE
Expose APIs for persisting RestProvider state 

### ISSUE THREAD

**Issue by sylvaingi** [coinbase/rest-hooks#69]
**Is your feature request related to a problem? Please describe.**
I'm exploring how to persist `<RestProvider />` state between reloads, using `localStorage`. Unfortunately this provider does not have a way to inform users its state has changed. I managed to get it working via `__INTERNAL__.StateContext`, but there should be a public API for this.
 
**Describe the solution you'd like**
Add a new `onStateChange` (or similar) prop to `<RestProvider/>`. Alongside `initialState` this new prop would allow full customization of the hydration & dehydration logic. 

Introducing a `cacheManager` prop with the following signature could work as well: 

```ts
interface CacheManager<T> {
  save: (state: State<T>) => void;
  load: () => State<T> | undefined;
}
```
Not sure how the `initialState` would behave in this case though.

**Describe alternatives you've considered**
I've successfully implemented this feature via the redux integration, but I feel like I shouldn't have to bring redux for this specific use case.

**Comment by ntucker** [comment]
Doing this via redux works by middleware, right? 2.0 rest hooks will allow arbitrary number of managers (which provide middleware to the reducer). If that works I may ask for an example to add to the docs 😉

**Comment by ntucker** [comment]
Refreshing myself with https://github.com/rt2zz/redux-persist

Most likely I would want to make this compatible with the ecosystem they have there - both transforms and storage engines. Seems like loading is generally considered something that should block render rather than trying to merge things - if this is the case that can easily be manged by simply passing initialState. Then the only thing left is how saving happens. Middleware only intercepts actions so there's definitely need to be a new integration point.

**Comment by sylvaingi** [comment]
Yep it works great with redux plus some persistence middleware, it just feels a bit overkill if I just want to add a persistence layer. One big advantage though is being able to inspect the cache via redux's dev tools.

Do you already have an idea how v2 managers will look like? 

Regarding redux-persist, I'm not a huge fan of the `<PersistGate>` thing, that's yet another component to insert at the top level of the app, and it's already quite crowded there between the routing stuff, suspense, providers and error boundaries. But they do support a lot of storage backends

> Middleware only intercepts actions so there's definitely need to be a new integration point.
Actually middlewares can access the store state, here's a quick and dirty middleware doing serialization to localStorage: https://github.com/sylvaingi/react-rest-hooks-test/blob/f9dc3faa8909da3a3a8f1ea302ef9e99f872f483/src/core/resources/cache/storage.ts#L7

**Comment by ntucker** [comment]
Manager plan: https://github.com/coinbase/rest-hooks/issues/15#issuecomment-480911714

Currently they just interact via middlewares...but if they can get the store maybe that's enough to work. However, since middlewares are run before reducers then it could be out of date. Tho I guess if you just schedule all callback you can get it after it commits.

You can see how the mangers are used here: https://github.com/coinbase/rest-hooks/blob/master/src/react-integration/provider/RestProvider.tsx#L26 They are just integrated into the reducer as middlewares. And then their cleanup is called on unmount.

Hmm, maybe I should start the 2.0 branch (was going to start this soon anyway) and you can see if a persist manager will work?

**Comment by sylvaingi** [comment]
Sure, that's something I can have a look at.

**Comment by sylvaingi** [comment]
Re: middlewares, if I'm not mistaken, at the end of the chain the actual dispatch is called, so calling `getState()` after should return the new state computed by the reducer, not the previous one. There would not be a need to defer to the next tick.

**Comment by ntucker** [comment]
Ah right....this brings me back 4 years....middleware acts as usual like an onion.

**Comment by ntucker** [comment]
https://github.com/coinbase/rest-hooks/tree/2.0/managers

Didn't update docs yet, but includes updated provider:

```tsx
interface ProviderProps {
    children: ReactNode;
    managers?: Manager[];
    initialState?: State<unknown>;
}
/** Controller managing state of the REST cache and coordinating network requests. */
export default function CacheProvider({ children, managers, initialState, }: ProviderProps): JSX.Element;
```

**Comment by ntucker** [comment]
BTW: if you want to turn this persist manager into a library, I'll gladly accept a docs PR with how-to-use guide.

**Comment by ntucker** [comment]
Just found a bug with the getState()....since it hadn't been used before: https://github.com/coinbase/rest-hooks/pull/70

Will need to add a test for this problem before I land.

**Comment by ntucker** [comment]
For some reason the tests are showing the same for before and after next() is called. But I had previously put a console.log() in my networkmanager's middleware and it was working - showing different things. 

See https://github.com/coinbase/rest-hooks/pull/70#discussion_r288355003

**Comment by ntucker** [comment]
Rebased https://github.com/coinbase/rest-hooks/tree/2.0/managers onto the state fix branch. See if it works for you since the Managers seem to work but the tests won't.

**Comment by sylvaingi** [comment]
Great, I'll start experimenting soon.

**Comment by ntucker** [comment]
Got the scheduler test working and landed in master. https://github.com/coinbase/rest-hooks/tree/2.0/managers is rebased on master with the latest

**Comment by ntucker** [comment]
Just rebased the https://github.com/coinbase/rest-hooks/tree/2.0/managers branch again

**Comment by ntucker** [comment]
Closing this since it should be possible to implement this using the now-released 2.0 API.

**Comment by ntucker** [comment]
> Re: middlewares, if I'm not mistaken, at the end of the chain the actual dispatch is called, so calling `getState()` after should return the new state computed by the reducer, not the previous one. There would not be a need to defer to the next tick.

https://github.com/coinbase/rest-hooks/pull/139 should fix this.
